### PR TITLE
NURBS patch attributes

### PR DESCRIPTION
**Pull request by dylan-copeland**
This PR adds functions to set and get attributes on NURBS patches and patch boundaries. The attributes are copied from the patches or patch boundaries to the mesh elements or boundary elements. There are also new functions in `NURBSExtension` to get the elements or boundary elements in each patch. A small example is given below, which can be tested on `data/pipe-nurbs.mesh`.
```
for (int i=0; i<mesh.NURBSext->GetNP(); ++i)
    mesh.SetPatchAttribute(i, 1000+i);
for (int i=0; i<mesh.NURBSext->GetNBP(); ++i)
    mesh.SetPatchBdrAttribute(i, 2000+i);
mesh.SetAttributes();
```

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3316](https://github.com/mfem/mfem/pull/3316) | @dylan-copeland | @tzanio | @tzanio + @v-dobrev + @kmittal2 | 11/13/22 | 3/30/23 | 3/30/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by JacobLotz** [comment]
Hi @dylan-copeland What would be the usage of these functions? As I correctly understand, the boundaries are usually set in the mesh. The purpose of this PR is to be able to hard wire it after reading your mesh? Such that it is the same for all meshes you read?

**Comment by JacobLotz** [comment]
I am afraid that at least one check is still failing.

File: mesh/nurbs.cpp [hunk 1]
```diff
@@ -3305,6 +3305,47 @@ void NURBSExtension::Set3DSolutionVector(Vector &coords, int vdim)
...
+
+   for (int e=0; e<bel_to_patch.Size(); ++e)
+   {
+      patch_to_bel[bel_to_patch[e]].Append(e);
+   }
+}
+
+const Array<int>& NURBSExtension::GetPatchElements(int patch)
+{
+   if (patch_to_el.size() == 0)
```
Comment (JacobLotz) [inline mesh/nurbs.cpp:3332]: This if-statement might slow down the code if this function is called often. I think a more efficient location to create `patch_to_el` might be after creating `el_to_patch` in `Generate2DElementDofTable()` for instance.  (As it inverts `el_to_patch` right?). Then as debug check a `MFEM_Assert` can be added here. @dylan-copeland have you looked into this?

File: mesh/mesh.cpp [hunk 2]
```diff
@@ -2774,6 +2774,41 @@ void Mesh::DoNodeReorder(DSTable *old_v_to_v, Table *old_elem_vert)
    Nodes->Update(); // just needed to update Nodes->sequence
 }
 
+void Mesh::SetPatchAttribute(int i, int attr)
+{
+   MFEM_ASSERT(NURBSext, "SetPatchAttribute is only for NURBS meshes");
+   NURBSext->SetPatchAttribute(i, attr);
+   const Array<int>& elems = NURBSext->GetPatchElements(i);
+   for (auto e : elems)
```
Comment (JacobLotz) [inline mesh/mesh.cpp:2782]: This kind of for loop with `for ( auto i: array)` might not be very consistent with a large part of the code in mesh.cpp. However, I see some applications in mesh.cpp and it forms a good alternative here, especially with the reference on line 2781.

File: mesh/mesh.cpp [hunk 3]
```diff
@@ -2774,6 +2774,41 @@ void Mesh::DoNodeReorder(DSTable *old_v_to_v, Table *old_elem_vert)
    Nodes->Update(); // just needed to update Nodes->sequence
 }
 
+void Mesh::SetPatchAttribute(int i, int attr)
+{
+   MFEM_ASSERT(NURBSext, "SetPatchAttribute is only for NURBS meshes");
+   NURBSext->SetPatchAttribute(i, attr);
```
Comment (JacobLotz) [inline mesh/mesh.cpp:2780]: I do not directly understand see why we have to set the attribute to patchTopo and the elements separately (line 2780 and 2784 resp). Is this for different usage? And, shouldn't these in the future be connected in some kind or does it have to do with the way NURBS are currently set up in MFEM?

File: mesh/mesh.cpp (hunk 2 above)
Comment (dylan-copeland) [reply mesh/mesh.cpp:2782]: This type of loop was introduced in C++11, but much of MFEM was written before C++11 was introduced in MFEM.

**Comment by dylan-copeland** [comment]
> Hi @dylan-copeland What would be the usage of these functions? As I correctly understand, the boundaries are usually set in the mesh. The purpose of this PR is to be able to hard wire it after reading your mesh? Such that it is the same for all meshes you read?

Thanks @JacobLotz for the review. For the code snippet in the PR desciption, if the mesh is printed via the `Print` function, the printed mesh file has the correct attributes for elements and boundary elements. In the `master` branch, the corresponding functions `SetAttribute` and `SetBdrAttribute`, can be called, but the resulting printed mesh will not have updated attributes. Instead, the original attributes will be used. This can be confusing to the user (e.g. myself). The reason is that `SetAttributes` only calls `elements[i]->SetAttribute(attr)`, and `Print` calls `Mesh::Printer`, which calls `NURBSExtension::Print`, which calls `patchTopo->PrintTopo`. Thus the attributes are set in `Mesh::elements` but not set in `patchTopo` for the `NURBSExtension` in `Mesh`. This PR copies the attributes from the former to the latter. This ensures that printed meshes have the correct attributes.

If the mesh generator sets the attributes as desired in the original mesh, then there is no need for this PR. This PR enables the user to modify attributes and print out a new NURBS mesh, with the new attributes.

File: mesh/nurbs.cpp (hunk 1 above)
Comment (dylan-copeland) [reply mesh/nurbs.cpp:3332]: I agree that this is a good suggestion, although I doubt that the if-statement may affect performance. Checking the size of an array should be faster than using the returned array, but in any case it makes sense to create `patch_to_el` only after creating `el_to_patch`. This change has been committed.

File: mesh/nurbs.hpp [hunk 4]
```diff
@@ -379,6 +385,14 @@ class NURBSExtension
    void GetVertexLocalToGlobal(Array<int> &lvert_vert);
    void GetElementLocalToGlobal(Array<int> &lelem_elem);
 
+   void SetPatchAttribute(int i, int attr) { patchTopo->SetAttribute(i, attr); }
```
[the remark this answers is not in the document]
Comment (dylan-copeland) [reply mesh/nurbs.hpp:390]: Documentation has been added.

**Comment by JacobLotz** [comment]
Hi @dylan-copeland , sorry for taking some time. Thank you for your clarifications! I think this PR is indeed very useful and we think this PR can be merged. I am looking forward to the reviews of other reviewers for their opinion.
Kind regards, Jacob

**Review by JacobLotz — APPROVED** [review]
Hi @dylan-copeland , sorry for taking some time. Thank you for your clarifications! I think this PR is indeed very useful and I think this PR can be merged. I am looking forward to the reviews of other reviewers for their opinion.
Kind regards, Jacob

File: mesh/nurbs.cpp [hunk 5]
```diff
@@ -3305,6 +3307,44 @@ void NURBSExtension::Set3DSolutionVector(Vector &coords, int vdim)
...
+
+   for (int e=0; e<bel_to_patch.Size(); ++e)
+   {
+      patch_to_bel[bel_to_patch[e]].Append(e);
+   }
+}
+
+const Array<int>& NURBSExtension::GetPatchElements(int patch)
+{
+   MFEM_ASSERT(patch_to_el.size() > 0, "patch_to_el not set");
```
Comment (kmittal2) [inline mesh/nurbs.cpp:3334]: I prefer if we do the same thing here that we do in `GetPatchBdrElements` by constructing the mapping if it does not already exists.

**Review by kmittal2 — APPROVED** [review]

**Review by IdoAkkerman — APPROVED** [review]
Sorry for being to late. But looks good!

**Comment by tzanio** [comment]
Thanks @dylan-copeland, should we list that in `CHANGELOG`?

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by dylan-copeland** [comment]
@tzanio CHANGELOG is updated.

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**